support allgather on GPU

janewangfb opened the pull request:
Summary: as titled

Differential Revision: D13266063

pietern commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -153,6 +153,32 @@ void initializeStreamsEvents(
+    }
+  }
+}
+
```
As discussed offline, we can probably allocate one stream per unique device index as well as just a single event per unique device index. There can only be 1 "current" stream per device, so as long as we synchronize with each one of them we're fine. CUDAStream is copyable so it can still be a nested vector of streams (it will just contain copies). CUDAEvent is NOT copyable, so it would require some other structure. We could do something like indirection where you have only a vector of streams and vector of events and then a nested vector of ints that correspond to the tensors, that index into the per-device streams and events vectors.

pietern commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -153,6 +153,28 @@ void initializeStreamsEvents(
   }
 }
 
+void initializeStreamsEvents(
```
Please add a comment similar to the one above with the of expectation that outputs in the inner vector are located on the same device.

janewangfb replied:
I reimplemented it and removed the assumption.

pietern commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -153,6 +153,28 @@ void initializeStreamsEvents(
+  streams.reserve(outputs.size());
+  events.resize(outputs.size());
+
+  for (size_t i = 0; i < outputs.size(); i++) {
```
Can you add a check here for the assumption that all outputs in outputs[i] are on the same device?

janewangfb replied:
the new code does not have this assumption

pietern commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -757,7 +779,9 @@ class AsyncAllgatherWork : public ProcessGroupGloo::AsyncWork {
   const uint32_t tag;
 
-  void run() override {
+  void allGather(
```
allgather in lowercase to match style in c10d.

janewangfb replied:
updated.

pietern commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -780,8 +804,90 @@ class AsyncAllgatherWork : public ProcessGroupGloo::AsyncWork {
+      device_guard.set_index(inputs[i].get_device());
+      AT_CUDA_CHECK(cudaStreamSynchronize(inputStreams[i]));
+    }
+    // Synchronize with copy operations.
```
There is no copy for the output streams so this comment can be removed.

janewangfb replied:
removed.

pietern reviewed (commented):
Some nits and one structural thing.

Right now there is an implied assumption that all outputs in the nested output vector are placed on the same device. We should test that this is the case and throw if it isn't. If some part of this functionality is reused and the assumption is false then is can lead to synchronization issues.

There is also the possibility of using `flattenDenseTensors` directly instead of first copying into temporary tensors on the CPU side and then flattening them. That would save another copy. Since this is a backfill op it is not critical but will result in improved performance. Can you file an issue to track this? It would make for a good starter task for somebody new to the code base.

pietern commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -128,31 +128,68 @@ at::Tensor pinnedLike(at::Tensor& tensor) {
+  at::cuda::OptionalCUDAGuard guard;
+  for (std::map<int64_t, size_t>::iterator it = streamMap.begin();
+       it != streamMap.end();
+       it++) {
```
You can use `for (const auto& it : streamMap)` for short (no need to change, just FYI)

janewangfb replied:
updated

pietern commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -128,31 +128,68 @@ at::Tensor pinnedLike(at::Tensor& tensor) {
+    // Ensure the new stream is synchronized with the current stream.
+    events[it->second].block(streams[it->second]);
+  }
+}
```
Can you add a comment at the top of this function describing what it does, and why? About the 1 stream/event per device, etc.

janewangfb replied:
comments added.

janewangfb replied:
added.

pietern reviewed (commented):
Looking good. There is a problem in CI though, something about invalid events.

pietern commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -128,31 +128,68 @@ at::Tensor pinnedLike(at::Tensor& tensor) {
+    std::vector<at::cuda::CUDAStream>& streams,
+    std::vector<at::cuda::CUDAEvent>& events) {
+  tensorStreamIndices.resize(tensors.size());
+  std::map<int64_t, size_t> streamMap;
```
Instead of int64_t, please use c10::DeviceIndex. Then to get it from the tensor, you can do tensor.device().index().

The get_device() function is going away (you couldn't have known this btw, I found out after investigating a bit since I've seen c10::Device and c10::DeviceIndex before).

janewangfb replied:
updated.

pietern commented:
Re: flattenDenseTensors, I think it would work with multiple devices, looking at the implementation: https://github.com/pytorch/pytorch/blob/7e4a5b89fe74ac3acff073075f67be4f8027d556/aten/src/THC/generic/THCTensorMath.cu#L87-L255

pietern commented:
Regarding DeviceIndex, also see #14729.

janewangfb commented:
crated https://github.com/pytorch/pytorch/issues/14812 for new comers.

janewangfb committed:
add gloo allgather support on GPU

Summary: as titled

Reviewed By: pietern

Differential Revision: D13266063

fbshipit-source-id: 413140d80df24f4d6db26d1fcb5051fc41b2ab9a
